drivers/input/input_tsc_keys.c

```diff
@@ -0,0 +1,174 @@
+#include <zephyr/sys/util.h>
+#include <zephyr/sys/ring_buffer.h>
+
+LOG_MODULE_REGISTER(tsc_keys, CONFIG_INPUT_LOG_LEVEL);
+
+struct input_tsc_keys_child_data {
+	uint32_t buffer[CONFIG_INPUT_TSC_KEYS_BUFFER_WORD_SIZE];
+	struct ring_buf rb;
+	enum {
+		INPUT_TSC_KEYS_EXPECT_DOWN,
+		INPUT_TSC_KEYS_EXPECT_UP,
+	} state;
+};
+
+struct input_tsc_keys_child_config {
+	char *label;
+	uint8_t group_index;
+	int32_t noise_threshold;
+	int zephyr_code;
+	struct input_tsc_keys_child_data *data;
+};
+
+struct input_tsc_keys_data {
+	struct k_timer sampling_timer;
+};
+
+struct input_tsc_keys_config {
+	const struct device *tsc_dev;
+	stm32_tsc_callback_t tsc_callback;
+	uint32_t sampling_interval_ms;
+	struct input_tsc_keys_child_config *children;
+	uint8_t child_num;
+};
+
+static void input_tsc_sampling_timer_callback(struct k_timer *timer)
+{
+	const struct device *dev = k_timer_user_data_get(timer);
+
+	stm32_tsc_start(dev);
+}
+
+static inline struct input_tsc_keys_child_config *
+input_tsc_child_config_get(const struct device *dev, uint8_t group)
+{
+	const struct input_tsc_keys_config *config = dev->config;
+
+	for (int i = 0; i < config->child_num; i++) {
+		const struct input_tsc_keys_child_config *child = &config->children[i];
+
+		if (child->group_index == group) {
+			return (struct input_tsc_keys_child_config *)child;
+		}
+	}
+
+	return NULL;
+}
+
+static void input_tsc_callback_handler(const struct device *dev, uint8_t group, uint32_t value)
+{
+	const struct input_tsc_keys_config *config = dev->config;
+
+	const struct input_tsc_keys_child_config *child_config =
+		input_tsc_child_config_get(dev, group);
+
+	if (!child_config) {
+		LOG_ERR("TSC@%p: No child config for group %d", config->tsc_dev, group);
+		return;
+	}
+
+	struct input_tsc_keys_child_data *child_data = child_config->data;
+
+	if (ring_buf_item_space_get(&child_data->rb) == 0) {
+		uint32_t oldest_point;
+		int32_t slope;
+
+		(void)ring_buf_get(&child_data->rb, (uint8_t *)&oldest_point, sizeof(oldest_point));
+
+		slope = value - oldest_point;
+		if (slope < -child_config->noise_threshold &&
+		    child_data->state == INPUT_TSC_KEYS_EXPECT_DOWN) {
+			/* this is a key press, now expect a release */
+			child_data->state = INPUT_TSC_KEYS_EXPECT_UP;
+			input_report_key(dev, child_config->zephyr_code, 1, false, K_NO_WAIT);
+		} else if (slope > child_config->noise_threshold &&
+			   child_data->state == INPUT_TSC_KEYS_EXPECT_UP) {
+			/* this is a key release, now expect a press */
+			child_data->state = INPUT_TSC_KEYS_EXPECT_DOWN;
+			input_report_key(dev, child_config->zephyr_code, 0, false, K_NO_WAIT);
+		}
+	}
+
+	(void)ring_buf_put(&child_data->rb, (uint8_t *)&value, sizeof(value));
+}
+
+int input_tsc_keys_init(const struct device *dev)
+{
+	const struct input_tsc_keys_config *config = dev->config;
+	struct input_tsc_keys_data *data = dev->data;
+
+	if (!device_is_ready(config->tsc_dev)) {
+		LOG_ERR("TSC@%p: TSC device not ready", config->tsc_dev);
```

Comment:
use `%s` and `config->tsc_dev.name`, then you have something that make sense without having to find address structures, same in all the other logs

Reply:
thanks for the tip